[SCEV] Introduce SCEVUse, use it instead of const SCEV * (NFCI).

fhahn opened the pull request:
This patch introduces SCEVUse, which is a tagged pointer containing the
used const SCEV *, plus extra bits to store NUW/NSW flags that are only
valid at the specific use.

This was suggested by @nikic as an alternative
to https://github.com/llvm/llvm-project/pull/90742.

This patch just updates most SCEV infrastructure to operate on SCEVUse
instead of const SCEV *. It does not introduce any code that makes use
of the use-specific flags yet which I'll share as follow-ups.

Note that this should be NFC, but currently there's at least one case
where it is not (turn-to-invariant.ll), which I'll investigate once we
agree on the overall direction.

This PR at the moment also contains a commit that updates various SCEV
clients to use `const SCEV *` instead of `const auto *`, to prepare for
this patch. This reduces the number of changes needed, as SCEVUse will
automatically convert to `const SCEV *`. This is a safe default, as it
just drops the use-specific flags for the expression (it will not drop
any use-specific flags for any of its operands though).

SCEVUse could probably also be used to address mis-compiles due to
equivalent AddRecs modulo flags result in an AddRec with incorrect flags
for some uses of some phis, e.g. the one
https://github.com/llvm/llvm-project/pull/80430 attempted to fix

Compile-time impact:
stage1-O3: +0.06%
stage1-ReleaseThinLTO: +0.07%
stage1-ReleaseLTO-g: +0.07%
stage2-O3: +0.11%

https://llvm-compile-time-tracker.com/compare.php?from=ce055843e2be9643bd58764783a7bb69f6db8c9a&to=8c7f4e9e154ebc4862c4e2716cedc3c688352d7c&stat=instructions:u

fhahn commented:
Example uses:

* Use SCEVUse to add additional NUW flag when reasoning in LAA: https://github.com/llvm/llvm-project/pull/91962
* get SCEV for Value with use-specific flags (only implemented for GEP exprs for now): https://github.com/llvm/llvm-project/pull/91964

fhahn commented:
ping :)

fhahn commented:
rebase & ping

fhahn commented:
Ping. Any thoughts about the overall direction? There are a number of places where context-specific SCEV flags could be helpful, with my main interest being on LAA at the moment.

fhahn commented:
Ping

fhahn commented:
Ping :)

Any thoughts on moving forward with this now we branched for the release?

nikic commented:
The first commit looks good, can you please land it?

nikic commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -69,6 +69,99 @@ enum SCEVTypes : unsigned short;
+  /// you should specify a real typed pointer or avoid this template.
+  ///
+  /// All clients should use assertions to do a run-time check to ensure that
+  /// this is actually true.
```
This copied comment doesn't make a lot of sense with a value of 0.

nikic commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -69,6 +69,99 @@ enum SCEVTypes : unsigned short;
+  }
+};
+
+template <typename To> [[nodiscard]] inline decltype(auto) dyn_cast(SCEVUse U) {
```
Is directly overloading dyn_cast the correct way to do this? Rather than providing a CastInfo implementation or something?

nikic commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -1363,8 +1458,9 @@ class ScalarEvolution {
-          C, detail::combineHashValue(reinterpret_cast<uintptr_t>(Op),
-                                      reinterpret_cast<uintptr_t>(Ty)));
+          C,
+          detail::combineHashValue(reinterpret_cast<uintptr_t>(Op.getPointer()),
```
Shouldn't this be getRawPointer()?

nikic reviewed (commented):
I assume that this patch is supposed to just mechanical changes for SCEVUse introduction -- otherwise this would need more adjustments, e.g. the poison analysis code would need to account for these flags, and these are proper poison flags rather than UB flags.

-----

I think my main high level question here is: What about SCEV unification? We get this straightforwardly for the top-level use flags because we can just compare the underlying pointers. But what about nested expressions?

If we have `((%a * %b)(u nuw) + %c)` and `((%a * %b) + %c)`, I assume we're actually going to treat these as two distinct, non-unified expressions. We currently rely on equality for SCEVs to be pointer equality, and this would somewhat break that. Do we need to have each SCEV also store a pointer to the unified SCEV without flags or something?

fhahn commented:
> I assume that this patch is supposed to just mechanical changes for SCEVUse introduction -- otherwise this would need more adjustments, e.g. the poison analysis code would need to account for these flags, and these are proper poison flags rather than UB flags.
> 

Yes exactly. The first real use would be in https://github.com/llvm/llvm-project/pull/91964

> I think my main high level question here is: What about SCEV unification? We get this straightforwardly for the top-level use flags because we can just compare the underlying pointers. But what about nested expressions?
> 
> If we have `((%a * %b)(u nuw) + %c)` and `((%a * %b) + %c)`, I assume we're actually going to treat these as two distinct, non-unified expressions. We currently rely on equality for SCEVs to be pointer equality, and this would somewhat break that. Do we need to have each SCEV also store a pointer to the unified SCEV without flags or something?

I've update the patch to make a distinction between canonical (no flags in any subexpression) and non-canonical. Only canonical SCEVs can be compared with `==`, with an assertion checking that. A canonical SCEV can be constructed by rebuilding the SCEV expression with all flags dropped. One of the bits in SCEVUse is used to indicate that the wrapped SCEV is canonical or not. So far it is not complete yet, and only supports expressions with flags in Add and Mul expressions.

This unfortunately about doubles the compile-time impact, worst increase is +0.18% for stage2-O3 ( https://llvm-compile-time-tracker.com/compare.php?from=9e831d50a0d67c6779048742657edb437818b681&to=d1a824d7f47b90faa4d708b543b78d28b761b18b&stat=instructions:u) 

I didn't spend any time on tuning yet, but there are likely more places that will need updating as we make use of more SCEVUses.

I added some basic tests for comparing SCEVs with and without flags here https://github.com/llvm/llvm-project/pull/91961/files#diff-218646cbfc7bf44fcc471db9ceb892a5a69a89edfa871a459cedddc991eb67a8



Adding a pointer to each SCEV would be an alternative, I originally thought tracking whether a SCEV is canonical using a bit in the pointer may be less overhead for code that doesn't use any flags, but I'll prototype it to see how it compares

fhahn reviewed (commented):
I also put together a version that stores the canonical SCEV via a pointer in each SCEV expression: https://github.com/llvm/llvm-project/commit/180f8be49b3d4db744704af11ad463a3f199d25a
 

Compile-time wise this seems to fare slightly better in most cases, with stage2-O3 being +0.17%
https://llvm-compile-time-tracker.com/compare.php?from=67c0846357bcd6faca713315380f9981a805a6e5&to=180f8be49b3d4db744704af11ad463a3f199d25a&stat=instructions:u

fhahn commented:
Ping :)


> I also put together a version that stores the canonical SCEV via a pointer in each SCEV expression: [180f8be](https://github.com/llvm/llvm-project/commit/180f8be49b3d4db744704af11ad463a3f199d25a)
> 
> Compile-time wise this seems to fare slightly better in most cases, with stage2-O3 being +0.17% https://llvm-compile-time-tracker.com/compare.php?from=67c0846357bcd6faca713315380f9981a805a6e5&to=180f8be49b3d4db744704af11ad463a3f199d25a&stat=instructions:u

WDYT would be the next steps to move forward with this, given the compile-time impact?

artagnon commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -65,6 +65,117 @@ enum SCEVTypes : unsigned short;
+  return CastInfo<To, const SCEV *>::isPossible(U.getPointer());
+}
+
+template <class X> auto dyn_cast_or_null(SCEVUse U) {
+  const SCEV *Val = U.getPointer();
+  if (!detail::isPresent(Val))
+    return CastInfo<X, const SCEV *>::castFailed();
+  return CastInfo<X, const SCEV *>::doCastIfPossible(detail::unwrapValue(Val));
+}
```
CastInfo definitely needs to be worked out, otherwise dyn_cast_if_present, cast_if_present etc. won't work.

fhahn replied:
Yep, will do that once we are happy with the rest of the patch

artagnon commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -65,6 +65,117 @@ enum SCEVTypes : unsigned short;
+
+/// Provide PointerLikeTypeTraits for SCEVUse, so it can be used with
+/// SmallPtrSet, among others.
+template <> struct PointerLikeTypeTraits<SCEVUse> {
```
I don't think there's a way to escape PointerLikeTraits, as SmallPtrSet needs to work.

artagnon commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -65,6 +65,117 @@ enum SCEVTypes : unsigned short;
+  const SCEV *getCanonical(ScalarEvolution &SE) {
+    if (isCanonical())
+      return getPointer();
+    return computeCanonical(SE);
```
I'm a little confused by why you can't computeCanonical at the callsites, and keep the canonical representation. Alternatively, why can't you computeCanonical in the constructor? Calling computeCanonical every time has caused a huge compile-time regression.

artagnon commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -65,6 +65,117 @@ enum SCEVTypes : unsigned short;
 
+class SCEV;
+
+class SCEVUse : public PointerIntPair<const SCEV *, 3> {
```
I think you need two bits: one for NUW and one for NSW. The canonical bit can be dropped, if you can canonicalize on construction.

artagnon reviewed (commented):
This is a very interesting patch! I can definitely see the value of having non-global SCEV flags, and I hope I was able to leave some comments that can help improve it.

fhahn commented:
Ping :) rebased after landing pattern matching and fixed the polly build failures.

Current CT impact: https://llvm-compile-time-tracker.com/compare.php?from=4e828f8d741ff61317bb1e0b67f22e274632b07a&to=f3905cdaf342e506e1e40dbef4d23b031dd83bd9&stat=instructions:u

artagnon commented:
Reverse ping. Do we have any viable alternative to resolve issues pertaining to the global wrap flags in SCEV? Several of them are quite complicated and exotic, but they result in real miscompiles. I think the compile-time regression is the cost to pay for a fundamental oversight in SCEV's design, and I think this patch pay for itself in terms of codegen improvements (although there's a long tail of follow-up patches required). Consider putting up an RFC for this patch, so that we get consensus on the compile-time cost, and get volunteers to help out with follow-ups if it lands? I think you can motivate it by scraping issues for long-standing miscompiles that it would eventually fix.

nikic reviewed (commented):
Would you mind rebasing this again?

I'll try to find some time to look at this soon(TM).

nikic reviewed (commented):
It might make sense to first just do a change that introduces a `SCEVUse` wrapper around `const SCEV *`, to get the mechanical change out of the way and avoid bitrot. The mass change to use SCEVUse makes it hard to spot the places where something more meaningful has changed.

fhahn committed:
[SCEV] Introduce SCEVUse wrapper type (NFC)

Add SCEVUse as a PointerIntPair wrapper around const SCEV * to prepare
for storing additional per-use information.

This commit contains the mechanical changes of adding an intial SCEVUse
wrapper and updating all relevant interfaces to take SCEVUse. Note that
currently the integer part is never set, and all SCEVUses are
considered canonical.

fhahn reviewed (commented):
More than happy to resume with this direction!

Rebased the PR and split up into 3 separate patches

1) add boilerplate SCEVUse update
2) Add CanonicalSCEV, always set to the created SCEV
3) Set canonical SCEV to potentially rewritten expression with flags dropped.

Will get compile-time latest compiletime numbers soon and look into some potential improvements to reduce overhead

fhahn reviewed (commented):
> I think you may have forgotten to push the new changes :)


ah yes, this is actually a user branch, should be updated now, but I still have to fix a few polly build failures I think

nikic commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -66,6 +66,115 @@ enum SCEVTypes : unsigned short;
+    return PointerLikeTypeTraits<SCEVUse>::getFromVoidPointer((void *)Val);
+  }
+
+  static unsigned getHashValue(SCEVUse U) { return hash_value(U); }
```
Just wondering where this hash_value() implementation is coming from, as I don't see it defined.

fhahn replied:
it was part of SCEVUse, but I moved it here.

nikic commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -66,6 +66,115 @@ enum SCEVTypes : unsigned short;
+
+  static inline SCEVUse getEmptyKey() {
+    uintptr_t Val = static_cast<uintptr_t>(-1);
+    Val <<= Log2MaxAlign;
```
I don't think this is really necessary in this context. We're putting data into the alignment bits anyway, so no need to go out of the way to preserve the alignment here, I think.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3759,7 +3775,9 @@ ScalarEvolution::getAddRecExpr(SmallVectorImpl<const SCEV *> &Operands,
   // with a SCEVCouldNotCompute as the cached BE count).
 
-  Flags = StrengthenNoWrapFlags(this, scAddRecExpr, Operands, Flags);
+  {
```
Why this block wrapper?

fhahn commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -4424,44 +4455,50 @@ ScalarEvolution::getSequentialMinMaxExpr(SCEVTypes Kind,
-  return Sequential ? getSequentialMinMaxExpr(scSequentialUMinExpr, Ops)
-                    : getMinMaxExpr(scUMinExpr, Ops);
+  if (Sequential) {
+    SmallVector<SCEVUse> SCEVOps(Ops.begin(), Ops.end());
```
was left over from the partial conversion, restored original code now, thanks

fhahn commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -2294,13 +2313,15 @@ CollectAddOperandsWithScales(SmallDenseMap<const SCEV *, APInt, 16> &M,
-                                       Add->operands(), NewScale, SE);
+        SmallVector<SCEVUse> Ops2;
+        for (const SCEV *S : Add->operands())
+          Ops2.emplace_back(S);
```
no longer needed now that all helpers have been convered to use SCEVUse, removed thanks

fhahn commented on llvm/lib/Transforms/Scalar/LoopDataPrefetch.cpp:
```diff
@@ -391,9 +391,10 @@ bool LoopDataPrefetch::runOnLoop(Loop *L) {
+    const SCEV *NextLSCEV = SE->getAddExpr(
+        P.LSCEVAddRec,
+        SE->getMulExpr(SE->getConstant(P.LSCEVAddRec->getType(), ItersAhead),
+                       P.LSCEVAddRec->getStepRecurrence(*SE).getPointer()));
```
not needed in the latest version, removed thanks!

fhahn commented on llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp:
```diff
@@ -1095,12 +1095,15 @@ void StraightLineStrengthReduce::allocateCandidatesAndFindBasisForGEP(
-    const SCEV *BaseExpr = SE->getGEPExpr(cast<GEPOperator>(GEP), IndexExprs);
+    SmallVector<SCEVUse, 4> IndexPtrs;
+    for (SCEVUse U : IndexExprs)
+      IndexPtrs.push_back(U);
```
No longer needed in the latest version, removed thanks!

fhahn commented on llvm/include/llvm/Analysis/ScalarEvolutionExpressions.h:
```diff
@@ -825,19 +830,19 @@ class SCEVRewriteVisitor : public SCEVVisitor<SC, const SCEV *> {
   }
 
   const SCEV *visitAddExpr(const SCEVAddExpr *Expr) {
```
Yep, maybe have a variant that passes the flags as additional arg, or something like that?

fhahn commented on llvm/include/llvm/Analysis/ScalarEvolutionExpressions.h:
```diff
@@ -964,15 +969,18 @@ class SCEVLoopAddRecRewriter
+    if (It == Map.end()) {
+      SmallVector<SCEVUse, 2> Operands2;
+      for (const SCEV *Op : Expr->operands())
+        Operands2.push_back(visit(Op));
```
no longer needed in latest version, removed thanks

fhahn committed:
!fixup address latest comments, thanks

nikic commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -959,21 +958,25 @@ static Immediate ExtractImmediate(const SCEV *&S, ScalarEvolution &SE) {
+    SmallVector<SCEVUse, 8> NewOps(Add->operands());
+    SCEVUse Back = NewOps.back();
+    GlobalValue *Result = ExtractSymbol(Back, SE);
+    NewOps.back() = Back;
```
Can probably revert this to the original form as well?

fhahn commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -2936,7 +2951,8 @@ const SCEV *ScalarEvolution::getAddExpr(SmallVectorImpl<const SCEV *> &Ops,
       SCEV::NoWrapFlags AddFlags = Flags;
       if (AddFlags != SCEV::FlagAnyWrap) {
-        auto *DefI = getDefiningScopeBound(LIOps);
+        SmallVector<SCEVUse, 8> LIOpPtrs(LIOps.begin(), LIOps.end());
```
no longer needed, removed thanks

fhahn commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3302,8 +3316,10 @@ const SCEV *ScalarEvolution::getMulExpr(SmallVectorImpl<const SCEV *> &Ops,
-    SmallVector<const SCEV *, 8> LIOps;
-    const SCEVAddRecExpr *AddRec = cast<SCEVAddRecExpr>(Ops[Idx]);
+    const SCEVAddRecExpr *AddRec = dyn_cast<SCEVAddRecExpr>(Ops[Idx]);
+    if (!AddRec)
```
fixed, probably messed up during the rebase, thanks

fhahn commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -4062,8 +4085,11 @@ class SCEVSequentialMinMaxDeduplicatingVisitor final
-    bool Changed = visit(Kind, NAry->operands(), NewOps);
+    SmallVector<SCEVUse, 8> Operands;
+    for (SCEVUse U : NAry->operands())
+      Operands.push_back(U.getPointer());
```
No longer needed, removed thanks

fhahn commented on llvm/lib/Transforms/Scalar/LoopFuse.cpp:
```diff
@@ -1135,9 +1135,10 @@ struct LoopFuser {
       if (ExprL == &OldL) {
-        append_range(Operands, Expr->operands());
+        for (SCEVUse Op : Expr->operands())
+          Operands.push_back(Op);
```
yep, restored, thanks

fhahn committed:
!fixup more cleanups

nikic reviewed (approved):
LGTM

fhahn committed:
!fixup address comments, thanks

nikic commented:
> > Before landing, just to recap on the compile-time impact of the change and follow-up changes needed
> 
> I assume we are fine with the impact and the general future direction/follow-up changes and will go ahead and merge this in a bit, unless I hear otherwise :)

Yes, I'm fine with the impact.

fhahn committed:
Merge branch 'main' into users/fhahn/scevuse

vzakhari commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -66,6 +66,76 @@ enum SCEVTypes : unsigned short;
+
+struct SCEVUse : PointerIntPair<const SCEV *, 2> {
+  SCEVUse() : PointerIntPair() { setFromOpaqueValue(nullptr); }
+  SCEVUse(const SCEV *S) : PointerIntPair() { setFromOpaqueValue((void *)S); }
```
It looks like this cast is going to break some builds:
```
error: cast from type ‘const llvm::SCEV*’ to type ‘void*’ casts away qualifiers [-Werror=cast-qual]
```
